[CodeGen][AArch64] Set min jump table entries to 13 for AArch64 targets

david-arm opened the pull request:
There are some workloads that are negatively impacted by using jump tables when the number of entries is small. The SPEC2017 perlbench benchmark is one example of this, where increasing the threshold to around 13 gives a ~1.5% improvement on neoverse-v1. I chose the minimum threshold based on empirical evidence rather than science, and just manually increased the threshold until I got the best performance without impacting other workloads. For neoverse-v1 I saw around ~0.2% improvement in the SPEC2017 integer geomean, and no overall change for neoverse-n1. If we find issues with this threshold later on we can always revisit this.

The most significant SPEC2017 score changes on neoverse-v1 were:

500.perlbench_r: +1.6%
520.omnetpp_r: +0.6%

and the rest saw changes < 0.5%.

I updated CodeGen/AArch64/min-jump-table.ll to reflect the new threshold. For most of the affected tests I manually set the min number of entries back to 4 on the RUN line because the tests seem to rely upon this behaviour.

david-arm committed:
[CodeGen][AArch64] Set min jump table entries to 13 for AArch64 targets

There are some workloads that are negatively impacted by using jump
tables when the number of entries is small. The SPEC2017 perlbench
benchmark is one example of this, where increasing the threshold to
around 13 gives a ~1.5% improvement on neoverse-v1. I chose the
minimum threshold based on empirical evidence rather than science,
and just manually increased the threshold until I got the best
performance without impacting other workloads. For neoverse-v1 I
saw around ~0.2% improvement in the SPEC2017 integer geomean, and
no overall change for neoverse-n1. If we find issues with this
threshold later on we can always revisit this.

The most significant SPEC2017 score changes on neoverse-v1 were:

500.perlbench_r: +1.6%
520.omnetpp_r: +0.6%

and the rest saw changes < 0.5%.

I updated CodeGen/AArch64/min-jump-table.ll to reflect the new
threshold. For most of the affected tests I manually set the min
number of entries back to 4 on the RUN line because the tests seem
to rely upon this behaviour.

davemgreen commented:
As with last time this came up I have no strong opinion on whether one value is better than another, so this sounds OK.

What happens with minsize? Should that have a lower limit?

david-arm committed:
Take minsize attribute into account

I've changed getMinimumJumpTableEntries to now take a Function
parameter so that we can query the minsize attribute and only
increase the threshold if minsize is absent.

paulwalker-arm reviewed (commented):
Now that you're making `MinimumJumpTableEntries` a per function property perhaps it’s better stored in `AArch64Subtarget` and then you don’t need to change the existing `getMinimumJumpTableEntries` interface.  This will be more in keeping with how other per function settings are handled.

david-arm committed:
Let the subtarget decide the minimum at construction

Instead of querying the function attributes every time, we can
let the subtarget decide once at construction.

david-arm commented:
> Now that you're making `MinimumJumpTableEntries` a per function property perhaps it’s better stored in `AArch64Subtarget` and then you don’t need to change the existing `getMinimumJumpTableEntries` interface. This will be more in keeping with how other per function settings are handled.

Done!

david-arm committed:
Fix clang format

davemgreen commented:
Thanks. It looks like some of the formatting could be redone? The failing test at least looks unrelated.

I see the Arm backend does the same thing with adding MinSize to the subtarget. This sounds OK to me if the formatting is happy.

david-arm committed:
Really fix the clang formatting issues

david-arm commented:
I ran make check-all using HEAD of LLVM and tests pass.

nikic commented:
How does this affect performance in cases where the last switch case is usually taken?

david-arm commented:
> How does this affect performance in cases where the last switch case is usually taken?

Honestly speaking I don't know! As mentioned in the commit message, I just changed this number until I got the best performance results, so it's quite possible that in some benchmarks that I've not tested this does make things worse. If we discover issues later on we can revisit this and perhaps refine the algorithm.

davemgreen commented:
> How does this affect performance in cases where the last switch case is usually taken?

If I understand what you are saying, it should generate a tree of branches so the positions of the cases shouldn't matter a huge amount. It can depend on the differences between the cases and the branch targets though, they might not match.

Let us know if you have reason to think this would be worse in some way, it's not a very easy thing to measure accurately. A load from a constant pool into an indirect jump can be worse than a series of branches that can each be independently branch predicted.

nikic commented:
> > How does this affect performance in cases where the last switch case is usually taken?
> 
> If I understand what you are saying, it should generate a tree of branches so the positions of the cases shouldn't matter a huge amount. It can depend on the differences between the cases and the branch targets though, they might not match.
> 
> Let us know if you have reason to think this would be worse in some way, it's not a very easy thing to measure accurately. A load from a constant pool into an indirect jump can be worse than a series of branches that can each be independently branch predicted.

I was mainly just wondering about what the worst case impact here would be for an unlucky switch. If I understand correctly, previously the switch (without jump table) would need at most two branches and now it needs at most four. Does that sound correct? (I was thinking in terms of a linear chain previously -- with a binary tree this doesn't sound so bad.)

sjoerdmeijer commented:
FWIW: this gives < 0.5% on the Neoverse V2 for Perlbench. Might even be noise because I haven't done a lot of runs.

sjoerdmeijer commented:
I think it's actually noise, don't think I see improvements on the V2. 
Perhaps that's still okay because it is shows improvements on the V1. But now I am actually wondering if the V1 improvement is an actual improvement, or do we get lucky with secondary effects such as different function/loop/block alignments?

david-arm commented:
> I think it's actually noise, don't think I see improvements on the V2. Perhaps that's still okay because it is shows improvements on the V1. But now I am actually wondering if the V1 improvement is an actual improvement, or do we get lucky with secondary effects such as different function/loop/block alignments?

I ran this quite a few times on neoverse-v1 hardware and consistently saw an improvement, particularly with https://github.com/llvm/llvm-project/pull/72132 applied as well. Although I appreciate there is a lot of noise in perlbench. From observation it seems that GCC also takes a similar approach when deciding to create jump tables for perlbench.

david-arm commented:
> I think it's actually noise, don't think I see improvements on the V2. Perhaps that's still okay because it is shows improvements on the V1. But now I am actually wondering if the V1 improvement is an actual improvement, or do we get lucky with secondary effects such as different function/loop/block alignments?

What I also saw was a gradual improvement in performance as you increased the threshold for creating jump tables. For example, I'd get consistently better results with a minimum threshold of 6, then even better with 10, then the best around 13. When I saw that trend of improving performance with higher threshold it made me think this wasn't just luck or coincidence, and that there is a genuine correlation.

sjoerdmeijer commented:
> > I think it's actually noise, don't think I see improvements on the V2. Perhaps that's still okay because it is shows improvements on the V1. But now I am actually wondering if the V1 improvement is an actual improvement, or do we get lucky with secondary effects such as different function/loop/block alignments?
> 
> What I also saw was a gradual improvement in performance as you increased the threshold for creating jump tables. For example, I'd get consistently better results with a minimum threshold of 6, then even better with 10, then the best around 13. When I saw that trend of improving performance with higher threshold it made me think this wasn't just luck or coincidence, and that there is a genuine correlation.

Ok, that sounds good.

Wilco1 commented:
> I think it's actually noise, don't think I see improvements on the V2. Perhaps that's still okay because it is shows improvements on the V1. But now I am actually wondering if the V1 improvement is an actual improvement, or do we get lucky with secondary effects such as different function/loop/block alignments?

Check the instruction count, cache and branch misses. Much of the gain comes from a huge reduction in instruction count, the rest from avoiding cache misses from loading the table entries and indirect branch misses. As indirect branch prediction improves, a lower value may be slightly better in future cores, however I don't believe it ever makes sense to use a table with less than 8 cases.

wangpc-pp commented on llvm/lib/Target/AArch64/AArch64Subtarget.cpp:
```diff
@@ -77,14 +77,19 @@ static cl::opt<AArch64PAuth::AuthCheckMethod>
                                         "to authenticated LR during tail call"),
                                cl::values(AUTH_CHECK_METHOD_CL_VALUES_LR));
 
+static cl::opt<unsigned> AArch64MinimumJumpTableEntries(
```
Do we still need another option for AArch64 here as we already have `-min-jump-table-entries` in `TargetLoweringBase.cpp`？

wangpc-pp replied:
OK, I get it. Sorry for the noise.
